File: server/src/repositories/BaseRepository.js
/**
 * Base Repository Pattern
 * 
 * Follows SOLID Principles:
 * - Single Responsibility: Handles only data access
 * - Open/Closed: Extendable without modification
 * - Liskov Substitution: All repositories can be substituted
 * - Dependency Inversion: Depends on abstractions (Model)
 */

class BaseRepository {
  constructor(Model) {
    this.Model = Model;
  }

  /**
   * Find by ID
   */
  async findById(id) {
    try {
      return await this.Model.findById(id);
    } catch (error) {
      throw new Error(`Failed to find ${this.Model.modelName} by ID: ${error.message}`);
    }
  }

  /**
   * Find one by query
   */
  async findOne(query) {
    try {
      return await this.Model.findOne(query);
    } catch (error) {
      throw new Error(`Failed to find ${this.Model.modelName}: ${error.message}`);
    }
  }

  /**
   * Find many by query
   */
  async findMany(query = {}, options = {}) {
    try {
      const { limit, skip, sort } = options;
      let queryBuilder = this.Model.find(query);

      if (skip) queryBuilder = queryBuilder.skip(skip);
      if (limit) queryBuilder = queryBuilder.limit(limit);
      if (sort) queryBuilder = queryBuilder.sort(sort);

      return await queryBuilder.exec();
    } catch (error) {
      throw new Error(`Failed to find ${this.Model.modelName} list: ${error.message}`);
    }
  }

  /**
   * Create new document
   */
  async create(data) {
    try {
      const document = new this.Model(data);
      return await document.save();
    } catch (error) {
      throw new Error(`Failed to create ${this.Model.modelName}: ${error.message}`);
    }
  }

  /**
   * Update by ID
   */
  async updateById(id, data, options = { new: true }) {
    try {
      return await this.Model.findByIdAndUpdate(id, data, options);
    } catch (error) {
      throw new Error(`Failed to update ${this.Model.modelName}: ${error.message}`);
    }
  }

  /**
   * Update one by query
   */
  async updateOne(query, data, options = { new: true }) {
    try {
      return await this.Model.findOneAndUpdate(query, data, options);
    } catch (error) {
      throw new Error(`Failed to update ${this.Model.modelName}: ${error.message}`);
    }
  }

  /**
   * Delete by ID
   */
  async deleteById(id) {
    try {
      return await this.Model.findByIdAndDelete(id);
    } catch (error) {
      throw new Error(`Failed to delete ${this.Model.modelName}: ${error.message}`);
    }
  }

  /**
   * Count documents
   */
  async count(query = {}) {
    try {
      return await this.Model.countDocuments(query);
    } catch (error) {
      throw new Error(`Failed to count ${this.Model.modelName}: ${error.message}`);
    }
  }

  /**
   * Check if document exists
   */
  async exists(query) {
    try {
      return await this.Model.exists(query);
    } catch (error) {
      throw new Error(`Failed to check ${this.Model.modelName} existence: ${error.message}`);
    }
  }
}

module.exports = BaseRepository;

File: lib/repositories/base_repository.dart
/// Base Repository Pattern
/// 
/// Follows SOLID Principles:
/// - Single Responsibility: Handles only data access
/// - Open/Closed: Extendable without modification
/// - Liskov Substitution: All repositories can be substituted
/// - Dependency Inversion: Depends on API abstractions

import '../api/base_api.dart';
import '../utils/api_cache.dart';
import '../utils/performance_utils.dart';

abstract class BaseRepository<T> {
  final BaseApi api;
  final ApiCache cache;
  final RequestDeduplicator deduplicator;

  BaseRepository({
    required this.api,
    ApiCache? cache,
    RequestDeduplicator? deduplicator,
  })  : cache = cache ?? ApiCache(),
        deduplicator = deduplicator ?? RequestDeduplicator();

  /// Convert JSON to model
  T fromJson(Map<String, dynamic> json);

  /// Convert model to JSON
  Map<String, dynamic> toJson(T model);

  /// Fetch with caching
  Future<T?> fetchWithCache(
    String cacheKey,
    Future<T> Function() fetchFunction, {
    bool useCache = true,
  }) async {
    // Check cache first
    if (useCache) {
      final cachedData = await cache.get(cacheKey);
      if (cachedData != null) {
        try {
          return fromJson(cachedData);
        } catch (e) {
          // Cache invalid, continue to fetch
        }
      }
    }

    // Fetch with deduplication
    return await deduplicator.deduplicate(cacheKey, () async {
      PerformanceMonitor.start(cacheKey);
      final result = await fetchFunction();
      PerformanceMonitor.end(cacheKey);

      // Cache the result
      await cache.put(cacheKey, toJson(result));

      return result;
    });
  }

  /// Fetch list with caching
  Future<List<T>> fetchListWithCache(
    String cacheKey,
    Future<List<T>> Function() fetchFunction, {
    bool useCache = true,
  }) async {
    // Check cache first
    if (useCache) {
      final cachedData = await cache.get(cacheKey);
      if (cachedData != null) {
        try {
          final List<dynamic> items = cachedData['items'] ?? [];
          return items.map((json) => fromJson(json as Map<String, dynamic>)).toList();
        } catch (e) {
          // Cache invalid, continue to fetch
        }
      }
    }

    // Fetch with deduplication
    return await deduplicator.deduplicate(cacheKey, () async {
      PerformanceMonitor.start(cacheKey);
      final result = await fetchFunction();
      PerformanceMonitor.end(cacheKey);

      // Cache the result
      await cache.put(cacheKey, {
        'items': result.map((item) => toJson(item)).toList(),
      });

      return result;
    });
  }

  /// Clear cache
  Future<void> clearCache(String cacheKey) async {
    await cache.remove(cacheKey);
  }
}

